docs/content/en/docs/Design/layout.md

```diff
@@ -0,0 +1,605 @@
+the shape of the domain tensor, and the range of a layout is always a 2D tensor
+whose first dimension denotes the ciphertext index and whose second index is the
+slot within that ciphertext.
+
+**Example 2:** Given a data matrix of type `tensor<8x8xi32>` and 8 ciphertexts
+with 32 slots each, the following layout implements the standard Halevi-Shoup
+diagonal layout.
+
+```
+{
+    [row, col] -> [ct, slot] :
+    0 <= row < 8
+    and 0 <= col < 8
+    and 0 <= ct < 8
+    and 0 <= slot < 32
+    and (row - col + ct) mod 8 = 0
+    and (row - slot) mod 32 = 0
+}
+```
+
+Note, this layout implements a diagonal packing, and further replicates each
+diagonal cyclically within a ciphertext.
+
+### Layout attributes
+
+Layouts are represented in HEIR via the `tensor_ext.layout` attribute. Its
+argument includes a string using the ISL notation above. For example
+
+```mlir
+#tensor_layout = #tensor_ext.layout<
+    "{ [i0] -> [ct, slot] : (slot - i0) mod 8 = 0 and ct = 0 and 1023 >= slot >= 0 and 7 >= i0 >= 0 }"
+>
+```
+
+Generally, layout attributes are associated with an SSA value by being attached
+to the op that owns the SSA value. In MLIR, which op owns the value has two
+cases:
+
+- For an op result, the layout attribute is stored on the op.
+- For a block argument, the layout attribute is stored on the op owning the
+  block, using the `OperandAndResultAttrInterface` to give a consistent API for
+  accessing the attribute.
+
+These two differences are handled properly by a helper library,
+`lib/Utils/AttributeUtils.h`, which exposes setters and getters for layout
+attributes. As of 2025-10-01, the system does not provide a way to handle ops
+with multiple regions or multi-block regions.
+
+For example:
+
+```mlir
+%1 = arith.addi %0, %1 {tensor_ext.layout = #layout_attr} : tensor<512xf32>
+```
+
+## Data-semantic and ciphertext-semantic tensors
+
+In HEIR, before lowering to scheme ops, we distinguish between types in two
+regimes:
+
+- _Data-semantic tensors_, which are scalars and tensors that represent
+  cleartext data values, largely unchanged from the original input program.
+- _Ciphertext-semantic tensors_, which are rank-2 tensors that represent packed
+  cleartext values in ciphertexts.
+
+The task of analyzing an IR to determine which layouts and kernels to use
+happens in the data-semantic regime. In these passes, chosen layouts are
+persisted between passes as attributes on ops (see
+[Layout attributes](#layout-attributes) above), and data types are unchanged.
+
+In this regime, there are three special `tensor_ext` ops that are no-ops on
+data-semantic type, but are designed to manipulate the layout attributes. These
+ops are:
+
+- `tensor_ext.assign_layout`, which takes a data-semantic value and a layout
+  attribute, and produces the same data-semantic type. This is an "entry point"
+  into the layout system and lowers to a loop that packs the data according to
+  the layout.
```

Comment:
```suggestion
  attribute, and produces the same data-semantic type. This is an "entry point"  into the layout system
  for cleartext values and lowers to a loop that packs the data according to the layout.
```

Actually I just wanted to understand this clearly - I was going to say that this is an "entry point" for values introduced into the IR as op results (and not block args) but then maybe cleartext block args are NOT assigned a layout.. (until they are used)?